File: tests/com/hotels/intellij/plugins/network/converter/FullHttpResponseToResponseContentConverterTest.java
package com.hotels.intellij.plugins.network.converter;

import io.netty.buffer.ByteBuf;
import io.netty.handler.codec.http.FullHttpResponse;
import io.netty.handler.codec.http.HttpHeaders;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class FullHttpResponseToResponseContentConverterTest {

    @Mock
    private DefaultByteBufToStringConverter defaultByteBufToStringConverter;

    @Mock
    private LZ4ByteBufToStringConverter lz4ByteBufToStringConverter;

    private FullHttpResponseToResponseContentConverter victim;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        victim = new FullHttpResponseToResponseContentConverter(defaultByteBufToStringConverter, lz4ByteBufToStringConverter);
    }

    @Test
    public void convertUsesDefaultConverterWhenContentTypeNotPresent() throws Exception {
        //given
        FullHttpResponse fullHttpResponse = mock(FullHttpResponse.class);
        HttpHeaders httpHeaders = mock(HttpHeaders.class);
        ByteBuf byteBuf = mock(ByteBuf.class);

        given(fullHttpResponse.headers()).willReturn(httpHeaders);
        given(fullHttpResponse.content()).willReturn(byteBuf);

        //when
        victim.convert(fullHttpResponse);

        //then
        verify(defaultByteBufToStringConverter).convert(byteBuf);
    }

    @Test
    public void convertUsesDefaultConverterWhenContentEncodingPresentButNotKnown() throws Exception {
        //given
        FullHttpResponse fullHttpResponse = mock(FullHttpResponse.class);
        HttpHeaders httpHeaders = mock(HttpHeaders.class);
        ByteBuf byteBuf = mock(ByteBuf.class);

        given(httpHeaders.get("Content-Encoding")).willReturn("unknown");
        given(fullHttpResponse.headers()).willReturn(httpHeaders);
        given(fullHttpResponse.content()).willReturn(byteBuf);

        //when
        victim.convert(fullHttpResponse);

        //then
        verify(defaultByteBufToStringConverter).convert(byteBuf);
    }

    @Test
    public void convertUsesLZ4ConverterWhenContentEncodingPresentAndKnown() throws Exception {
        //given
        FullHttpResponse fullHttpResponse = mock(FullHttpResponse.class);
        HttpHeaders httpHeaders = mock(HttpHeaders.class);
        ByteBuf byteBuf = mock(ByteBuf.class);

        given(httpHeaders.get("Content-Encoding")).willReturn("lz4-block");
        given(fullHttpResponse.headers()).willReturn(httpHeaders);
        given(fullHttpResponse.content()).willReturn(byteBuf);

        //when
        victim.convert(fullHttpResponse);

        //then
        verify(lz4ByteBufToStringConverter).convert(byteBuf);
    }

    @Test
    public void convertPrettyPrintsContentWhenContentTypeJsonAndValidJson() throws Exception {
        //given
        FullHttpResponse fullHttpResponse = mock(FullHttpResponse.class);
        HttpHeaders httpHeaders = mock(HttpHeaders.class);
        ByteBuf byteBuf = mock(ByteBuf.class);

        given(httpHeaders.get("Content-Type")).willReturn("application/json");
        given(defaultByteBufToStringConverter.convert(byteBuf)).willReturn("{\"valid\": \"json\"}");
        given(fullHttpResponse.headers()).willReturn(httpHeaders);
        given(fullHttpResponse.content()).willReturn(byteBuf);

        //when
        String content = victim.convert(fullHttpResponse);

        //then
        assertThat(content, is("{\n  \"valid\": \"json\"\n}"));
    }

    @Test
    public void convertDoesNotPrettyPrintContentWhenContentTypeJsonAndInvalidJson() throws Exception {
        //given
        FullHttpResponse fullHttpResponse = mock(FullHttpResponse.class);
        HttpHeaders httpHeaders = mock(HttpHeaders.class);
        ByteBuf byteBuf = mock(ByteBuf.class);

        given(httpHeaders.get("Content-Type")).willReturn("application/json");
        given(defaultByteBufToStringConverter.convert(byteBuf)).willReturn("{\"invalid\": \"json\"");
        given(fullHttpResponse.headers()).willReturn(httpHeaders);
        given(fullHttpResponse.content()).willReturn(byteBuf);

        //when
        String content = victim.convert(fullHttpResponse);

        //then
        assertThat(content, is("{\"invalid\": \"json\""));
    }
}

File: src/test/kotlin/com/hotels/intellij/plugins/network/converter/FullHttpResponseToResponseContentConverterTest.kt
package com.hotels.intellij.plugins.network.converter

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import io.netty.buffer.ByteBuf
import io.netty.handler.codec.http.FullHttpResponse
import io.netty.handler.codec.http.HttpHeaders
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test

class FullHttpResponseToResponseContentConverterTest {

    private val defaultByteBufToStringConverter = mockk<DefaultByteBufToStringConverter>()
    private val lz4ByteBufToStringConverter = mockk<LZ4ByteBufToStringConverter>()

    private var victim = FullHttpResponseToResponseContentConverter(defaultByteBufToStringConverter, lz4ByteBufToStringConverter)

    @Test
    @Throws(Exception::class)
    fun convertUsesDefaultConverterWhenContentTypeNotPresent() {
        //given
        val fullHttpResponse = mockk<FullHttpResponse>()
        val httpHeaders = mockk<HttpHeaders>()
        val byteBuf = mockk<ByteBuf>()
        every { httpHeaders["Content-Encoding"] } returns null
        every { httpHeaders["Content-Type"] } returns null
        every { fullHttpResponse.headers() } returns httpHeaders
        every { fullHttpResponse.content() } returns byteBuf

        every { defaultByteBufToStringConverter.convert(byteBuf) } returns ""

        //when
        victim.convert(fullHttpResponse)

        //then
        verify { defaultByteBufToStringConverter.convert(byteBuf) }
    }

    @Test
    @Throws(Exception::class)
    fun convertUsesDefaultConverterWhenContentEncodingPresentButNotKnown() {
        //given
        val fullHttpResponse = mockk<FullHttpResponse>()
        val httpHeaders = mockk<HttpHeaders>()
        val byteBuf = mockk<ByteBuf>()
        every { httpHeaders["Content-Encoding"] } returns "unknown"
        every { httpHeaders["Content-Type"] } returns null
        every { fullHttpResponse.headers() } returns httpHeaders
        every { fullHttpResponse.content() } returns byteBuf

        every { defaultByteBufToStringConverter.convert(byteBuf) } returns ""

        //when
        victim.convert(fullHttpResponse)

        //then
        verify { defaultByteBufToStringConverter.convert(byteBuf) }
    }

    @Test
    @Throws(Exception::class)
    fun convertUsesLZ4ConverterWhenContentEncodingPresentAndKnown() {
        //given
        val fullHttpResponse = mockk<FullHttpResponse>()
        val httpHeaders = mockk<HttpHeaders>()
        val byteBuf = mockk<ByteBuf>()
        every { httpHeaders["Content-Encoding"] } returns "lz4-block"
        every { httpHeaders["Content-Type"] } returns null
        every { fullHttpResponse.headers() } returns httpHeaders
        every { fullHttpResponse.content() } returns byteBuf

        every { lz4ByteBufToStringConverter.convert(byteBuf) } returns ""

        //when
        victim.convert(fullHttpResponse)

        //then
        verify { lz4ByteBufToStringConverter.convert(byteBuf) }
    }

    @Test
    @Throws(Exception::class)
    fun convertPrettyPrintsContentWhenContentTypeJsonAndValidJson() {
        //given
        val fullHttpResponse = mockk<FullHttpResponse>()
        val httpHeaders = mockk<HttpHeaders>()
        val byteBuf = mockk<ByteBuf>()
        every { httpHeaders["Content-Encoding"] } returns null
        every { httpHeaders["Content-Type"] } returns "application/json"
        every { fullHttpResponse.headers() } returns httpHeaders
        every { fullHttpResponse.content() } returns byteBuf

        every { defaultByteBufToStringConverter.convert(byteBuf) } returns "{\"valid\": \"json\"}"

        //when
        val content = victim.convert(fullHttpResponse)

        //then
        assertThat(content, `is`("{\n  \"valid\": \"json\"\n}"))
    }

    @Test
    @Throws(Exception::class)
    fun convertDoesNotPrettyPrintContentWhenContentTypeJsonAndInvalidJson() {
        //given
        val fullHttpResponse = mockk<FullHttpResponse>()
        val httpHeaders = mockk<HttpHeaders>()
        val byteBuf = mockk<ByteBuf>()
        every { httpHeaders["Content-Encoding"] } returns null
        every { httpHeaders["Content-Type"] } returns "application/json"
        every { fullHttpResponse.headers() } returns httpHeaders
        every { fullHttpResponse.content() } returns byteBuf

        every { defaultByteBufToStringConverter.convert(byteBuf) } returns "{\"invalid\": \"json\""

        //when
        val content = victim.convert(fullHttpResponse)

        //then
        assertThat(content, `is`("{\"invalid\": \"json\""))
    }
}